### PR TITLE
Adding openSUSE Tumbleweed to meson build

### PR DESCRIPTION
**Pull request by nbyavuz**
Able to add openSUSE Tumbleweed to CI, all tests are passing. 
Related pipeline run: https://cirrus-ci.com/build/5949148653420544

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -225,6 +225,99 @@ task:
...
+    - lsblk
+    - ulimit -a -H
+    - ulimit -a -S
+    - export
+  sysconfig_script:
+    - useradd -m -U postgres
+    - chown -R postgres:postgres .
+    - mkdir -p /tmp/ccache_dir
+    - chown -R postgres:postgres /tmp/ccache_dir
+    - echo '* - nofile 65536' > /etc/security/limits.d/postgres.conf
```
Comment (anarazel) [inline .cirrus.yml:254]: Hm - why is this needed?

File: .cirrus.yml [hunk 2]
```diff
@@ -225,6 +225,99 @@ task:
...
+    - su postgres -c 'ulimit -l -H'
+    - su postgres -c 'ulimit -l -S'
+    # - echo '/tmp/%e-%s-%p.core' > /proc/sys/kernel/core_pattern
```
Comment (anarazel) [inline .cirrus.yml:257]: Does core dump handling actually work without this?

File: ci/docker/linux_opensuse_tumbleweed [hunk 3]
```diff
@@ -0,0 +1,19 @@
+FROM opensuse/tumbleweed
+RUN \
+  zypper -n update && \
+  zypper -n install \
+  shadow util-linux util-linux-systemd systemd-devel \
```
Comment (anarazel) [inline ci/docker/linux_opensuse_tumbleweed:5]: This seems to have failed quite a few times: https://cirrus-ci.com/task/5207362771353600 
Do you have any idea why that is? Perhaps there's a better mirror?

File: ci/docker/linux_opensuse_tumbleweed [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  git make gcc gcc-c++ perl-IPC-Run \
+  ccache clang meson ninja bison flex diff gettext-runtime \
+  llvm-devel icu libicu-devel liblz4-devel libopenssl-devel \
+  readline-devel pam-devel tcl-devel python3-devel \
+  libxml2-devel libxslt-devel libldapcpp-devel libuuid-devel \
+  krb5-devel krb5-client krb5-server krb5-plugin-kdb-ldap \
+  openldap2 openldap2-client
```
Comment (anarazel) [inline ci/docker/linux_opensuse_tumbleweed:12]: For the debian container I now put these into one-package-per-line and grouped them into different sections. See https://github.com/anarazel/pg-vm-images/blob/containers/scripts/linux_debian_install_deps.sh
Perhaps worth copying?

File: ci/docker/linux_opensuse_tumbleweed [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+# Fixing libperl.so not found error
+RUN \
+  find /usr/lib/ -name libperl.so -exec ln -sf '{}' /usr/lib64/libperl.so ';'
+# Fixing ICU errors caused by locale being set to 'POSIX' 
+ENV LANG en_US.UTF-8
+ENV LANGUAGE en_US:en
+ENV LC_ALL en_US.UTF-8
```
Comment (anarazel) [inline ci/docker/linux_opensuse_tumbleweed:19]: Hm, odd that that's required. Is it even for the autoconf build?

The different ENV lines should probably be combined so that they result in one docker layer instead of three...

File: ci/docker/linux_opensuse_tumbleweed [hunk 6]
```diff
@@ -0,0 +1,19 @@
+FROM opensuse/tumbleweed
+RUN \
+  zypper -n update && \
+  zypper -n install \
+  shadow util-linux util-linux-systemd systemd-devel \
+  git make gcc gcc-c++ perl-IPC-Run \
+  ccache clang meson ninja bison flex diff gettext-runtime \
```
Comment (anarazel) [inline ci/docker/linux_opensuse_tumbleweed:7]: https://cirrus-ci.com/task/5207362771353600?logs=build#L26 says
> 'diff' not found in package names. Trying capabilities.

Seems worth just using diffutils instead.

File: ci/docker/linux_opensuse_tumbleweed (hunk 3 above)
Comment (anarazel) [inline ci/docker/linux_opensuse_tumbleweed:5]: Why do we need util-linux-systemd?

File: ci/docker/linux_opensuse_tumbleweed (hunk 4 above)
Comment (anarazel) [inline ci/docker/linux_opensuse_tumbleweed:12]: Should probably clean zypper's cache, so that all that stuff doesn't end up in the container.

File: ci/docker/linux_opensuse_tumbleweed (hunk 5 above)
Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: No, autoconf was working correctly but meson was failing with
`ERROR:  collations with different collate and ctype values are not supported by ICU
CONTEXT:  SQL statement "CREATE COLLATION test1 (provider = icu, lc_collate = 'C', lc_ctype = 'C.UTF-8');"
PL/pgSQL function inline_code_block line 3 at EXECUTE` error while running `pg_regress/collate.linux.utf8` test.

File: ci/docker/linux_opensuse_tumbleweed (hunk 4 above)
Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: Done.

File: .cirrus.yml (hunk 1 above)
Comment (nbyavuz) [reply .cirrus.yml]: You are right, not needed, removed.

File: ci/docker/linux_opensuse_tumbleweed (hunk 6 above)
Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: Done.

File: ci/docker/linux_opensuse_tumbleweed (hunk 3 above)
Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: Otherwise it fails when it run `lsblk` command.

Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: I am not sure about that, it didn't give errors in any other run. I added `zypper ref` for refreshing mirrors, just in case.

File: ci/docker/linux_opensuse_tumbleweed (hunk 4 above)
Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: Done.

File: .cirrus.yml (hunk 2 above)
Comment (nbyavuz) [reply .cirrus.yml:256]: It doesn't work. For it to work, docker needs to run on privileged mode and for docker to run on privileged mode, It should run on a VM. See: https://cirrus-ci.org/guide/docker-builder-vm/

File: ci/docker/linux_opensuse_tumbleweed (hunk 3 above)
Comment (anarazel) [reply ci/docker/linux_opensuse_tumbleweed]: I think it'd be fine to just remove that...

Comment (nbyavuz) [reply ci/docker/linux_opensuse_tumbleweed]: Done.

**Comment by anarazel** [comment]
Could you rebase these changes? They don't apply anymore, and the diff is pretty hard to review now :(

**Comment by nbyavuz** [comment]
> Could you rebase these changes? They don't apply anymore, and the diff is pretty hard to review now :(

Done.

> Able to add openSUSE Tumbleweed to CI, all tests are passing. Related pipeline run: https://cirrus-ci.com/build/5949148653420544



> Could you rebase these changes? They don't apply anymore, and the diff is pretty hard to review now :(

Done.

**Comment by nbyavuz** [comment]
> Could you rebase these changes? They don't apply anymore, and the diff is pretty hard to review now :(

Done.

**Comment by anarazel** [comment]
Could you rebase once more? Then I'll merge.

**Comment by nbyavuz** [comment]
> Could you rebase once more? Then I'll merge.

Done.

File: .cirrus.yml [hunk 7]
```diff
@@ -18,6 +18,8 @@ env:
 
   CCACHE_MAXSIZE: "250M"
 
+  CI_FOLDER: src/tools/ci
+
```
Comment (anarazel) [inline .cirrus.yml:22]: This won't change anymore, so it can be removed now.

File: .cirrus.yml [hunk 8]
```diff
@@ -212,14 +250,152 @@ task:
...
+    chown root:postgres /tmp/cores
+
+
+task:
+  <<: *linux_opensuse_template
+
+  name: Linux - OpenSuse Tumbleweed - Autoconf
+
+  configure_script: |
+    su postgres <<-EOF
```
Comment (anarazel) [inline .cirrus.yml:335]: There's a trailing whitespace here

File: .cirrus.yml [hunk 9]
```diff
@@ -212,14 +250,152 @@ task:
...
+      --with-python \
+      --with-ssl=openssl \
+      --with-systemd \
+      --with-tcl --with-tclconfig=/usr/lib64/ \
+      --with-uuid=e2fs \
+      \
+      CC="ccache gcc"  \
+      CXX="ccache g++" \
+      CLANG="ccache clang" \
+      CFLAGS="-O0 -ggdb"
```
Comment (anarazel) [inline .cirrus.yml:359]: Another trailing whitespace.

File: .cirrus.yml [hunk 10]
```diff
@@ -212,14 +250,152 @@ task:
...
+
+  test_world_script: |
+    su postgres <<-EOF
+      ulimit -c unlimited; \
+      make -s ${CHECK} ${CHECKFLAGS} -j${TEST_JOBS}
+    EOF
+
+  on_failure:
+    <<: *on_failure
+    cores_script: ${CI_FOLDER}/cores_backtrace.sh linux /tmp/cores
```
Comment (anarazel) [inline .cirrus.yml:372]: Probably no point in setting up the core script for now, it won't work as-is, due to not being able to set the core pattern.

File: .cirrus.yml [hunk 11]
```diff
@@ -212,14 +250,152 @@ task:
...
+
+task:
```
Comment (anarazel) [inline .cirrus.yml:374]: nitpick: There's two newlines between other tasks.

**Review by anarazel — APPROVED** [review]
I left a few last minute minor changes. Could you address them, rebase, squash your changes into a single commit, and then merge?

File: .cirrus.yml (hunk 7 above)
Comment (nbyavuz) [reply .cirrus.yml]: Done.

File: .cirrus.yml (hunk 8 above)
Comment (nbyavuz) [reply .cirrus.yml]: Done.

File: .cirrus.yml (hunk 9 above)
Comment (nbyavuz) [reply .cirrus.yml]: Done.

File: .cirrus.yml (hunk 10 above)
Comment (nbyavuz) [reply .cirrus.yml]: Yes, removed from both.

File: .cirrus.yml (hunk 11 above)
Comment (nbyavuz) [reply .cirrus.yml:372]: Done.

**Comment by nbyavuz** [comment]
> I left a few last minute requests for minor changes. Could you address them, rebase, squash your changes into a single commit, and then merge?

Done.

File: src/backend/main/main.c [hunk 12]
```diff
@@ -30,6 +30,10 @@
 #include <sys/param.h>
 #endif
 
+#if defined(WIN32)
+#include <crtdbg.h>
+#endif
+
```
Comment (anarazel) [inline src/backend/main/main.c]: Why is this commit still here? Seems like a mismerge?

File: .cirrus.yml [hunk 13]
```diff
@@ -396,11 +440,68 @@ task:
...
+
+      tests_script:
+        - ulimit -c unlimited
+        - ulimit -n 1024
+        - meson test --no-rebuild -C build
+
+  always:
+    meson_log_artifacts:
+      path: "build/meson-logs/*.txt"
+      type: text/plain
```
Comment (anarazel) [inline .cirrus.yml]: None of these changes should still be here, I think?

File: .cirrus.yml [hunk 14]
```diff
@@ -242,53 +242,11 @@ task:
...
-
-      configure_script:
-        - su postgres -c 'meson setup --buildtype debug -Dcassert=true -Dssl=openssl -Duuid=ossp build'
-      build_script:
-        - su postgres -c 'ninja -C build'
-      upload_caches:
-        - ccache
-
-      tests_script:
-        - su postgres -c 'meson test --no-rebuild -C build'
```
Comment (anarazel) [inline .cirrus.yml]: Re changes shouldn't be here, in a prior comment: Note that you're removing the changes you made in that prior commit here.

**Review by anarazel — COMMENTED** [review]
If you want I can do the rebase? I think there's enough merge conflict resolutions in here that the easiest thing would be to just squash the three commits into one, and then look at removing the few remaining unnecessary changes

**Comment by anarazel** [comment]
Wohoo!
